File: Group D/Code/python/IMGsharp.py
import cv2
import os

source_directory = "../opencv_dataset/Blurry"
destination_directory = "IMGSharpDone"
os.makedirs(destination_directory, exist_ok=True)

start_time = cv2.getTickCount()

laplacian_kernel = 3

for filename in os.listdir(source_directory):
    if filename.lower().endswith(('.jpg', '.jpeg', '.png', '.bmp', '.gif')):
        input_image = cv2.imread(os.path.join(source_directory, filename))

        if input_image is not None:
            channels = cv2.split(input_image)
            sharpened_channels = []

            for channel in channels:
                sharp_channel = cv2.Laplacian(channel, cv2.CV_16S, ksize=laplacian_kernel)
                sharp_channel = cv2.convertScaleAbs(sharp_channel)
                sharpened_channels.append(sharp_channel)

            sharpened_image = cv2.merge(sharpened_channels)

            destination_path = os.path.join(destination_directory, filename)
            cv2.imwrite(destination_path, sharpened_image)

            print(f"Image harpened and saved: {destination_path}")

end_time = cv2.getTickCount()
execution_time = (end_time - start_time) / cv2.getTickFrequency()

print(f"Image sharpening. Sharpened images saved in '{destination_directory}'.")
print(f"Execution Time: {execution_time} seconds")

File: Group D/Code/rust/IMGsharp.rs
use opencv::{imgcodecs, imgproc, core, types};
use std::{env, fs};

#[derive(Debug)]
enum CustomError {
    OpenCVError(opencv::Error),
    IOError(std::io::Error),
}

impl From<opencv::Error> for CustomError {
    fn from(error: opencv::Error) -> Self {
        CustomError::OpenCVError(error)
    }
}

impl From<std::io::Error> for CustomError {
    fn from(error: std::io::Error) -> Self {
        CustomError::IOError(error)
    }
}

fn main() -> Result<(), CustomError> {
    if let Ok(cwd) = env::current_dir() {
        println!("Current working directory: {:?}", cwd);
    } else {
        eprintln!("Failed to get the current working directory.");
    }

    let input_folder = "../../../../opencv_dataset/Blurry";
    let output_folder = "../../IMGsharpDone";

    fs::create_dir_all(output_folder)?;

    let entries = fs::read_dir(input_folder)?;
    let tick_frequency = core::get_tick_frequency()?;
    let start_time = core::get_tick_count()?;

    for entry in entries {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() {
            if let Some(ext) = path.extension() {
                if ext == "jpg" || ext == "jpeg" {
                    if let Ok(image) = imgcodecs::imread(path.to_str().unwrap(), imgcodecs::IMREAD_COLOR) {
                        let kernel = core::Mat::from_slice(&[0.0, -1.0, 0.0, -1.0, 5.0, -1.0, 0.0, -1.0, 0.0])?;

                        let mut sharpened_image = core::Mat::default();
                        imgproc::filter_2d(&image, &mut sharpened_image, core::CV_16S, &kernel, core::Point::new(-1, -1), 0.0, core::BORDER_DEFAULT)?;

                        let mut converted_image = core::Mat::default();
                        core::convert_scale_abs(&sharpened_image, &mut converted_image, 1.0, 0.0)?;

                        let file_name = path.file_name().unwrap().to_string_lossy();
                        let output_path = format!("{}/{}", output_folder, file_name);

                        imgcodecs::imwrite(&output_path, &converted_image, &types::VectorOfi32::new())?;

                        println!("Image sharpened and saved:: {}", output_path);
                    }
                }
            }
        }
    }

    let end_time = core::get_tick_count()?;
    let execution_time = ((end_time as f64) - (start_time as f64)) / tick_frequency;

    println!("Execution Time: {:.6} seconds", execution_time);

    Ok(())
}
